### PR TITLE
Draft of <coroutine>, not yet built.

### PR DESCRIPTION
**Pull request by StephanTLavavej**
Please note that acceptance of community PRs will be delayed while we are
bringing our test and CI systems online. For more information, see the
README.md.

# Description
Implements #40. This will need to be developed in conjunction with changes to the compiler front-end.

# Checklist:

- [x] I understand README.md.
- [x] If this is a feature addition, that feature has been voted into the C++
  Working Draft.
- [x] Any code files edited have been processed by clang-format 8.0.1.
  (The version is important because clang-format's behavior sometimes changes.)
- [x] Identifiers in any product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 .
- [ ] Identifiers in test code changes are *not* `_Ugly`. **(no tests yet)**
- [ ] Test code includes the correct headers as per the Standard, not just
  what happens to compile. **(no tests yet)**
- [ ] The STL builds and test harnesses have passed (must be manually verified
  by an STL maintainer before CI is online, leave this unchecked for initial
  submission). **(neither built nor tested yet)**
- [x] This change introduces no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate or
  trivially copyable, etc.). If unsure, leave this box unchecked and ask a
  maintainer for help.

### CODE REVIEW

File: stl/inc/coroutine [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+// * coroutine_handle<>::operator() and coroutine_handle<>::resume:
+//   Expects: `*this` refers to a coroutine suspended at other than its final suspend point.
+//   Effects: Resumes the execution of the coroutine.
+// * coroutine_handle<>::destroy: s/Requires/Expects/
+// * coroutine_handle<Promise>::promise: s/Requires/Expects/
+
+// intrinsics used in the implementation of coroutine_handle
+_EXTERN_C
+size_t _coro_resume(void*); // FIXME: Calling convention?
+#pragma intrinsic(_coro_resume) // FIXME: Is `#pragma intrinsic` necessary?
```
Comment (sylveon) [inline stl/inc/coroutine]: > Some intrinsics are available only as intrinsics, and some are available both in function and intrinsic implementations. You can instruct the compiler to use the intrinsic implementation in one of two ways, depending on whether you want to enable only specific functions or you want to enable all intrinsics. The first way is to use `#pragma intrinsic(intrinsic-function-name-list)`.
https://docs.microsoft.com/en-us/cpp/intrinsics/compiler-intrinsics

If there's no method implementation, I believe `#pragma intrinsic` is optional`.

File: stl/inc/coroutine [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    constexpr explicit operator bool() const noexcept {
+        return _Ptr != nullptr;
+    }
+
+    _NODISCARD bool done() const noexcept { // strengthened
+        if (reinterpret_cast<intptr_t>(_Ptr) == _Noop_sentinel_value) {
+            return false;
+        }
+
+        // _STL_ASSERT(_Ptr != nullptr);
```
Comment (CaseyCarter) [inline stl/inc/coroutine]: Reviewer note: All of these `_STL_ASSERT`s are commented out because I intended this to be a `(core)` header. Our headers marked `(core)` don't require or cause linking with the library. Consequently, the assertion failure handler might not available.

I suppose with today's technology we could uncomment these and make them `_STL_INTERNAL_CHECK`s.

File: stl/inc/coroutine [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+//     A program shall not declare an explicit or partial specialization of `coroutine_handle`.
+// * coroutine_handle<Promise>::from_promise:
+//   Expects: `p` references the promise object of a coroutine.
+//   Returns: A coroutine handle `h` that refers to the coroutine whose promise object is referenced by `p`. [Note:
+//            `addressof(h.promise()) == addressof(p)` is consequently `true`.-end note]
+//   Throws: Nothing.
+// * coroutine_handle<>::address:
+//   Returns: A non-null pointer value if and only if `*this` refers to a coroutine.
+// * coroutine_handle<>::from_address and coroutine_handle<Promise>::from_address:
+//   Expects: ...
```
Comment (CaseyCarter) [inline stl/inc/coroutine]: More LWG issue fodder: "*Requires*: `addr` was obtained via a prior call to `address`." is insufficient for `coroutine_handle<Promise>::from_address`; it must require that `addr` is a null pointer value or was obtained via `h.address()` for a coroutine handle `h` that refers to a coroutine whose promise object has type `Promise`.

**Comment by jblazquez** [comment]
This is very exciting work!

Not sure if this is a library or compiler change, but will [symmetric coroutine control transfer](http://open-std.org/JTC1/SC22/WG21/docs/papers/2018/p0913r0.html) work after these changes are merged?

**Comment by StephanTLavavej** [comment]
> will symmetric coroutine control transfer work

I am a coroutine novice, but perhaps @CaseyCarter can answer this.

**Comment by CaseyCarter** [comment]
> Not sure if this is a library or compiler change, but will [symmetric coroutine control transfer](http://open-std.org/JTC1/SC22/WG21/docs/papers/2018/p0913r0.html) work after these changes are merged? In particular the part about returning a coroutine from await_suspend.

Symmetric control transfer is purely a compiler feature which MSVC doesn't yet fully implement. The compiler does perform a symmetric transfer as of VS 2019 16.3 if you return a `coroutine_handle` from `await_suspend`, but it doesn't yet know to "clean up" the prior coroutine's stack frame before transferring control to the new one. This is fine for trivial/demonstration programs, but will overflow the stack quickly in long-running async code that performs lots of symmetric transfers.

TLDR: Don't do this in production code.

@GorNishanov can correct me if my understanding is out of date.

**Comment by GorNishanov** [comment]
Yes. The frontend will accept the source code that uses symmetric control transfer:

```c++
#include <future>
using namespace std;
using namespace std::experimental;

struct S {
  bool await_ready() { return false; }
  void await_resume() {}
  // Look symmetric control transfer compiles!
  auto await_suspend(coroutine_handle<> h) { return h;  }
};

// WARNING: future-adapter is not in C++20. 
// But acceptable for a quick demo.
future<void> f() {
  co_await S{};
  co_return;
}

int main() {
  puts("Hello");
  f().get();
  getchar();
}
```

But, if you look at the call stack, you will see that it results in a normal call, not symmetric control transfer, because the backend work is not finished yet.

	QuickTest.exe!f$_ResumeCoro$2() Line 16	C++ <<<--- will disappear once BE work is done.
 	QuickTest.exe!f$_ResumeCoro$2() Line 15	C++
 	QuickTest.exe!f$_InitCoro$1() Line 17	C++
 	QuickTest.exe!f() Line 17	C++
 	QuickTest.exe!main() Line 21	C++

File: stl/inc/coroutine [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+struct _Resumable_frame {
+    using _Fn = void __cdecl(void*);
+
+    _Fn* _Resume;
+    unsigned short _Index;
+    unsigned short _Flags;
+};
+
+inline constexpr intptr_t _Noop_sentinel_value = -1;
```
Comment (miscco) [inline stl/inc/coroutine]: You are never using the value directly, but always in conjunction with a 'reinterpret_cast' would it be possible to directly define
```cpp
inline constexpr void* _Noop_sentinel_pointer = reinterpret_cast<void*>(intptr_t{-1});
```

Comment (miscco) [reply stl/inc/coroutine]: Derp, `reinterpret_cast` is not constexpr. Given that this is essentially a fixed bit-pattern why not `static_cast`

Comment (sylveon) [reply stl/inc/coroutine]: static_cast cannot convert integer types to pointer types

Comment (StephanTLavavej) [reply stl/inc/coroutine]: Yep!

**Comment by StephanTLavavej** [comment]
My coworkers have persuaded me that there's no need to keep this draft PR open for months on end. We'll be working with the compiler front-end team on this feature in the coming months and will share a more-finished version at that time.
